### PR TITLE
Master dnd block nby

### PR DESCRIPTION
**Pull request by Goaman**
- [REF] web_editor,mass_mailing,website: make `closestElement` support `predicate`
- [REF] web_editor: prepare absolute containers for the movable node 
- [IMP] web_editor: drag and drop nodes within the editable

PURPOSE
The main reason is that when there are two block side by side that are
not editable (eg. the `HR` tag, the `/template` block), it is impossible
to set the cursor in between to add more content.

task-2984709

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+        };
+        // This container help to add zone into which the mouse can activate the move widget.
+        this._widgetHookContainer = this._editor.makeAbsoluteContainer('oe-widget-hooks-container', positionOption);
+        // This container contains the differents widgets.
+        this._widgetContainer = this._editor.makeAbsoluteContainer('oe-widgets-container', positionOption);
+        // This container contains the jquery helper element.
+        this._dragHelperContainer = this._editor.makeAbsoluteContainer('oe-movenode-helper-container', positionOption);
+        // This container contains drop zones. They are the zones that handle where the drop should happen.
+        this._dropzonesContainer = this._editor.makeAbsoluteContainer('oe-dropzones-container', positionOption);
+        // This container contains drop hint. The final rectangle showed to the user.
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:53]: Thanks for those comments about the purpose of the different containers ! 👍

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+        // this._widgetHookContainer.classList.add('debug');
+        // this._dropzonesContainer.classList.add('debug');
+
+        this._scrollableElement = this._editable.parentElement;
+        while (this._scrollableElement) {
+            if (getComputedStyle(this._scrollableElement).overflowY === 'auto') {
+                break;
+            }
+            this._scrollableElement = this._scrollableElement.parentElement;
+        }
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:66]: It looks like this is a pattern that we use a lot. Maybe we need some kind of `ancestor` util ?
```js
function ancestor(node, predicate) {
    do {
        node = node?.parentElement();
    } while (node && predicate && !predicate(ancestor))
    return node;
}
```

... but then you'd need to have a look around in the code to find other places where we use this pattern to use the helper function instead.

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+        this._moveWidget.style.height = `${widgetMoveSize}px`;
+        this._moveWidget.style.top = `${anchorY - containerRect.y - moveWidgetOffsetTop}px`;
+        this._moveWidget.style.left = `${anchorX - containerRect.x - widgetContainerWidth}px`;
+        const timeout = setTimeout(()=> {
+            return this._moveWidget.classList.add('oe-sidewidget-move--show');
+        });
+
+        // Keep a reference for the closure.
+        const moveWidget = this._moveWidget;
+        this._removeMoveWidget = () => {
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:194]: I don't understand why we need to programmatically write a new method on `this`. If I understand correctly, it's to be sure to call it at some points ? Like basically everywhere we call it it's like "oh, right, if we had this thing going on, we need to cancel it", right ? I guess it's somewhat ok then... The first alternative I see would be to have a lot of variables on `this` to store the information that you already have here thanks to scoping so it wouldn't be better. Another alternative I see is to have a reference on `this` that stores all those callbacks, something like `_cancellableMoveWidgets` and you'd have a (haha again) one-time method `cancelMoveWidgets` which would cycle on those and call them all, but I guess you can only have one at a time anyway so it would be overkill. So, for those reasons, I think this is probably a viable way to do it.

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+            this._editor.enableAvatars(movableElement);
+            moveWidget.classList.remove('oe-sidewidget-move--show');
+            const transitionend = () => {
+                moveWidget.removeEventListener('transitionend', transitionend);
+                moveWidget.remove();
+            };
+            moveWidget.addEventListener('transitionend', transitionend);
+
+            // Ensure the widget is removed even if the the transition does not end.
+            setTimeout(() => moveWidget.remove(), 200);
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:205]: 👀

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+            this._removeMoveWidget = undefined;
+            this._currentMovableElement = undefined;
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:207]: Oh but you _do_ have this kind of variable _anyway_... In fact it looks like the only values that you're missing to have `_removeMoveWidget` be a proper method of the class is the avatars stuff...

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+        this._moveWidget.style.width = `${widgetMoveSize}px`;
+        this._moveWidget.style.height = `${widgetMoveSize}px`;
+        this._moveWidget.style.top = `${anchorY - containerRect.y - moveWidgetOffsetTop}px`;
+        this._moveWidget.style.left = `${anchorX - containerRect.x - widgetContainerWidth}px`;
+        const timeout = setTimeout(()=> {
+            return this._moveWidget.classList.add('oe-sidewidget-move--show');
+        });
+
+        // Keep a reference for the closure.
+        const moveWidget = this._moveWidget;
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:193]: Wait, does that mean you can have two of those `_removeMoveWidget` callbacks at the same time ? If not then why do you need to keep the reference since the next time you're going to set it you'd have called `_removeMoveWidget` anyway, and if yes then what prevents you from overriding one `this._removeMoveWidget` with another ? I'm very confused.

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+            moveWidgetOffsetTop = parseInt(style.marginTop, 10) || 0;
+        }
+
+        this._moveWidget.style.width = `${widgetMoveSize}px`;
+        this._moveWidget.style.height = `${widgetMoveSize}px`;
+        this._moveWidget.style.top = `${anchorY - containerRect.y - moveWidgetOffsetTop}px`;
+        this._moveWidget.style.left = `${anchorX - containerRect.x - widgetContainerWidth}px`;
+        const timeout = setTimeout(()=> {
+            return this._moveWidget.classList.add('oe-sidewidget-move--show');
+        });
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:190]: 👀

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+        this.observer = new MutationObserver(() => {
+            this._removeMoveWidget && this._removeMoveWidget();
+            this._updateHooks();
+        });
+        this.observer.observe(this._editable, {
+            childList: true,
+            subtree: true,
+            characterData: true,
+            characterDataOldValue: true,
+        });
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:79]: I don't get it... you _observe_ every change in the editor for that ? Isn't that a bit overkill ? I mean you only see the handle on your current block anyway. I feel like I missing some important piece of understanding here.

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+            // setTimeout as the mutation observer will remove it otherwise.
+            setTimeout(() => this._setMovableElement(movableElement));
+        }
+    }
+    _onMousemove(e) {
+        this._updateAnchorWidgets(e.target);
+    }
+    _onDocumentKeydown() {
+        this._removeMoveWidget && this._removeMoveWidget();
+    }
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:368]: I'm confused. What is the use case for document keydown ?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 10]
```diff
@@ -208,7 +209,7 @@ const Wysiwyg = Widget.extend({
             },
             commands: powerboxOptions.commands,
             categories: powerboxOptions.categories,
-            plugins: options.editorPlugins,
+            plugins: [...(options.editorPlugins || []), [MoveNodePlugin, { wysiwyg: this }]],
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:212]: This `wysiwyg` option doesn't seem to be used. Did I miss something ? Because if it's not used then we might remove support for plugin options... To be honest this was not even an option per se, it was basically misusing an options mechanism to pass some reference around... ^^'

**Review by dmo-odoo — COMMENTED** [review]
Lots of comments here but it's a great feature that requires quite a bit of code so that makes sense. 👍 There are comments that are more "discussions" like things where I'm not sure I understood some details properly and that we might want to discuss together at GR.

File: addons/note/static/src/scss/note.scss [hunk 11]
```diff
@@ -31,7 +31,7 @@
         }
         .o_field_html .note-editable {
             border: none;
-            padding: $o-sheet-vpadding $o-horizontal-padding 10px !important;
+            padding: $o-sheet-vpadding 30px 10px !important;
```
[the remark this answers is not in the document]
Comment (Goaman) [reply addons/note/static/src/scss/note.scss:34]: 16px

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 12]
```diff
@@ -737,11 +742,24 @@ export class OdooEditor extends EventTarget {
      * @param {*} className A class to add to the container in order to make
      *              the container more visible in the devtool and potentially
      *              add css rules for the container and it's children.
+     * @param {string}  options.position How the container should be added.
+     * @param {string}  options.targetContainerId An existing container from
+     *              which we add the created container depending.
      */
-    makeAbsoluteContainer(className) {
+    makeAbsoluteContainer(className, { position = 'append', targetContainerId = undefined } = {}) {
```
[the remark this answers is not in the document]
Comment (Goaman) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:749]: - the reason: more easily identify the container in the devtools
- to be able to target it when creating another absolute container

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 13]
```diff
@@ -737,11 +742,24 @@ export class OdooEditor extends EventTarget {
...
         const container = this.document.createElement('div');
         container.className = `oe-absolute-container ${className}`;
-        this._mainAbsoluteContainer.append(container);
+        let targetElement = targetContainerId && this._mainAbsoluteContainer.querySelector(`.oe-absolute-container.${targetContainerId}`);
```
[the remark this answers is not in the document]
Comment (Goaman) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:752]: To be discussed irl.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/commands.js [hunk 14]
```diff
@@ -800,10 +795,7 @@ export const editorCommands = {
                     const column = columns.pop();
                     const columnContents = unwrapContents(column);
                     for (const node of columnContents) {
-                        node.ouid = undefined; // Allow move out of unbreakable
-                        for (const descendant of descendants(node)) {
-                            descendant.ouid = undefined; // Allow move out of unbreakable
-                        }
+                        resetUnbreakable(node);
```
[the remark this answers is not in the document]
Comment (Goaman) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/commands/commands.js:798]: `freeOuids`, `resetOuids`, `resetUnbreakableIds` ?

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js (hunk 2 above)
Comment (Goaman) [reply addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:66]: Problem with our closestElement util: it limit the node to be in the `odoo-editor-editable`. In this case, the node is outside the editable.

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js (hunk 8 above)
Comment (Goaman) [reply addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:79]: Perfs looks good on my machine. Between 2 and 3ms for `_updateHooks`.

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js (hunk 9 above)
Comment (Goaman) [reply addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:357]: comment added

File: addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js (hunk 2 above)
Comment (Goaman) [reply addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js:55]: updated to use `closestElement`

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (Goaman) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:847]: As we talked in person, I've renamed the attribute `className` into `containerId`

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (Goaman) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:850]: I've removed options for the method.

**Comment by Goaman** [comment]
@dmo-odoo ready for re-review

**Comment by dmo-odoo** [comment]
@Goaman It is not clear to me what you expect me to review. All the commits with a non-standard name ?

**Comment by Goaman** [comment]
@dmo-odoo I've squashed the commit for reviewing all the commits.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 15]
```diff
@@ -2930,6 +2930,13 @@ export class OdooEditor extends EventTarget {
         }
 
         const sel = this.document.getSelection();
+        if (!sel.rangeCount || !closestElement(sel.focusNode) || !closestElement(sel.anchorNode)) {
+            if (this.options.autohideToolbar) {
+                this.toolbar.style.visibility = 'hidden';
+            }
+            return;
+        }
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2969]: This is a separate fix that feels out of place in this commit.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 16]
```diff
@@ -285,22 +285,24 @@ export function getFurthestUneditableParent(node, parentLimit) {
...
-export function closestElement(node, selector) {
-    let element = node;
-    while (element && element.nodeType !== Node.ELEMENT_NODE) {
-        element = element.parentNode;
-    }
-    if (element && selector) {
-        element = element.closest(selector);
-    }
-    return element && element.querySelector('.odoo-editor-editable') ? null : element;
+export function closestElement(node, { selector = "*", predicate = () => true } = {}) {
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js]: It feels so clumsy to have to write `{ selector: ...}` all the time we use this function, which is a lot. I suppose you'll disagree but personally I think I'd rather have a polymorphic parameter which is either a string or a function. In typescript it would be like `predicate: string | (node: Node) => Boolean`. This commit would become 10 lines instead of hundreds, and the enterprise PR could even be deleted.

File: addons/web_editor/static/src/js/editor/smooth_scroll_on_drag.js [hunk 17]
```diff
@@ -1,13 +1,9 @@
...
 
-import Class from "@web/legacy/js/core/class";
-import mixins from "@web/legacy/js/core/mixins";
-
 /**
  * Provides a helper for SmoothScrollOnDrag options.offsetElements
  */
-const OffsetElementsHelper = Class.extend({
-
+class OffsetElementsHelper {
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/smooth_scroll_on_drag.js:6]: Shouldn't this be in its own commit ?

Comment (Goaman) [reply addons/web_editor/static/src/js/editor/smooth_scroll_on_drag.js:6]: done

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/powerbox.test.js [hunk 18]
```diff
@@ -64,6 +64,7 @@ describe('Powerbox', () => {
     describe('class', () => {
         it('should properly order default commands and categories', async () => {
             const editable = document.createElement('div');
+            editable.classList.add('odoo-editor-editable');
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/test/spec/powerbox.test.js:67]: I don't remember why we needed to add this ?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 19]
```diff
@@ -285,22 +285,27 @@ export function getFurthestUneditableParent(node, parentLimit) {
...
- * @returns {HTMLElement}
- */
-export function closestElement(node, selector) {
-    let element = node;
-    while (element && element.nodeType !== Node.ELEMENT_NODE) {
-        element = element.parentNode;
-    }
-    if (element && selector) {
-        element = element.closest(selector);
+ * @param {string | Function} [selector='*']
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:294]: Rename to `selector`.

**Comment by dmo-odoo** [comment]
@robodoo rebase-ff r+
